Enable use of Unix domain sockets

bblancha opened the pull request:
* Use traits to hide socket type from server and client loops
* Preserve tcp-based api, don't break current clients

bblancha committed:
Enable use of Unix domain sockets

* Use traits to hide socket type from server and client loops
* Preserve tcp-based api, don't break current clients

bblancha commented:
I still need to check if this builds on Windows.
[edit] Now it compiles on Windows.

stepancheg commented on src/client.rs:
```diff
@@ -36,25 +36,60 @@ use client_conf::*;
+{
     pub event_loop: Option<reactor::Remote>,
-    pub addr: Option<SocketAddr>,
+    pub addr: Option<T>,
```
I think probably instead of type parameters field should be of enum type like

```
enum AnySocketAddr { Inet(...), Unix(...) }
```

Because generic builder leads to bad caller ergonomics.

stepancheg commented on tests/test_misc/server_test_unix.rs:
```diff
@@ -0,0 +1,66 @@
+
+struct Blocks {}
+
+impl Service for Blocks {
```
This looks like copy-paste.

bblancha replied:
Yes, I copy-pasted the server_test.rs file, then made small changes to make a UDS ServerTest struct..
I will refactor them back together and clean up a little bit.

stepancheg reviewed (changes requested):
Apart from minor changed I requested, it looks good.

bblancha committed:
Address review comments

* Don't duplicate code in server_test
* Use enum to hold either unix address or inet address
    - server.local_addr() now returns an enum.
    - server.local_addr.port() will now return a result<i16>
        that contains a port number if an inet socket is being used.

bblancha commented:
Still need to change interface of ClientBuilder to match ServerBuilder

bblancha committed:
Refactor ClientBuilder to avoid generics

stepancheg commented on tests/server.rs:
```diff
@@ -391,6 +396,22 @@ pub fn http_1_1() {
+pub fn http_1_1_unix() {
+    env_logger::init().ok();
+
+    let _server = ServerTest::new_unix("/tmp/rust_http2_test".to_owned());
```
Fixed path in `/tmp` is not ideal. Something like [that](https://github.com/rust-lang-nursery/tempdir) could be used.

stepancheg commented on tests/test_misc/server_test.rs:
```diff
@@ -22,7 +22,6 @@ pub struct ServerTest {
     pub port: u16,
 }
 
-
```
This change needs to be reverted.

stepancheg reviewed (approved):
There are minor issues with commit, but I'm going to squash and merge it now anyway.

stepancheg commented:
Do I understand correctly that grpc implementation based on this patch should be compatible with other grpc implementations over Unix sockets?

bblancha commented:
Sounds good, I will follow up with fixes in a few days.

bblancha commented:
Yes, I believe it will be compatible. I haven't tested interop yet, but will do so soon.
